okhttp/src/jvmMain/kotlin/okhttp3/internal/connection/ConnectPlan.kt

```diff
@@ -132,6 +132,7 @@ class ConnectPlan(
       success = true
       return ConnectResult(plan = this)
     } catch (e: IOException) {
+      client.routeDatabase.failed(route)
```

Comment:
This will also capture cases where the call was canceled. Hmm.

Reply:
And looking at the other call to routeDatabase.failed(), I think we should notify the proxy selector.

This is from `RealConnection.kt`:

```
    // Tell the proxy selector when we fail to connect on a fresh connection.
    if (failedRoute.proxy.type() != Proxy.Type.DIRECT) {
      val address = failedRoute.address
      address.proxySelector.connectFailed(
        address.url.toUri(), failedRoute.proxy.address(), failure
      )
    }
```

Perhaps the nicest structure is to move that snippet out of `RealConnection` and into `RouteDatabase`, so the two are always in sync.

Reply:
Yep, I wanted this to prompt discussion.  I'm not clear that the current exception filtering is obviously right.

I'll split it out and use the opportunity to make a test.